controllers/pod_controller.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/appuio/scheduler-canary-controller/controllers/podstate"
+)
+
+const (
+	instanceLabel     = "scheduler-canary-controller.appuio.io/instance"
+	timeoutAnnotation = "scheduler-canary-controller.appuio.io/timeout"
+)
+
+// PodReconciler reconciles a Pod object
+type PodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reconciles Pods with the `instanceLabel` set.
+// It tracks state changes in an annotation on the pod and calculates the timing metrics on pod completion or timeout.
+// Completed pods, pods with no timeout annotation, and timed out pods are deleted.
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	pod := &corev1.Pod{}
+	err := r.Client.Get(ctx, req.NamespacedName, pod)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	if !pod.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Object is in the process of being deleted.
+		return ctrl.Result{}, nil
+	}
+
+	instance := pod.GetLabels()[instanceLabel]
+
+	timeout, err := timeoutFromPod(pod)
+	if err != nil {
+		l.Error(err, "Pod has no valid timeout annotation, deleting immediately")
+	}
```

Comment:
Nitpick: In this case you rely on `timeout` being 0 in an error case. I'm generally not a huge fan of using the returned value in error cases.

Also I don't think `ParseDuration` gives you the guarantee to always return 0 if there is an error. The implementation does, but it could technically change that (It realistically never will though so :shrug: )

Reply:
I'll explicitly set timeout to zero on error case. Also to make it easier to read and understand.